core/imds/src/main/java/software/amazon/awssdk/imds/internal/DefaultEc2MetadataAsyncClient.java

```diff
@@ -0,0 +1,310 @@
+
+import java.net.URI;
+import java.nio.ByteBuffer;
+import java.time.Duration;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import org.reactivestreams.Publisher;
+import software.amazon.awssdk.annotations.Immutable;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.annotations.ThreadSafe;
+import software.amazon.awssdk.core.exception.RetryableException;
+import software.amazon.awssdk.core.exception.SdkClientException;
+import software.amazon.awssdk.core.http.HttpResponseHandler;
+import software.amazon.awssdk.core.interceptor.ExecutionAttributes;
+import software.amazon.awssdk.core.internal.http.TransformingAsyncResponseHandler;
+import software.amazon.awssdk.core.internal.http.async.AsyncResponseHandler;
+import software.amazon.awssdk.core.internal.http.async.SimpleHttpContentPublisher;
+import software.amazon.awssdk.core.retry.RetryPolicyContext;
+import software.amazon.awssdk.http.HttpStatusFamily;
+import software.amazon.awssdk.http.SdkHttpFullRequest;
+import software.amazon.awssdk.http.SdkHttpFullResponse;
+import software.amazon.awssdk.http.SdkHttpResponse;
+import software.amazon.awssdk.http.async.AsyncExecuteRequest;
+import software.amazon.awssdk.http.async.SdkAsyncHttpClient;
+import software.amazon.awssdk.http.async.SdkHttpContentPublisher;
+import software.amazon.awssdk.http.crt.AwsCrtAsyncHttpClient;
+import software.amazon.awssdk.imds.Ec2MetadataAsyncClient;
+import software.amazon.awssdk.imds.Ec2MetadataRetryPolicy;
+import software.amazon.awssdk.imds.EndpointMode;
+import software.amazon.awssdk.imds.MetadataResponse;
+import software.amazon.awssdk.utils.IoUtils;
+import software.amazon.awssdk.utils.Logger;
+
+@SdkInternalApi
+@Immutable
+@ThreadSafe
+public final class DefaultEc2MetadataAsyncClient implements Ec2MetadataAsyncClient {
+
+    private static final Logger log = Logger.loggerFor(DefaultEc2MetadataClient.class);
+    private static final int DEFAULT_RETRY_THREAD_POOL_SIZE = 3;
+
+    private final Ec2MetadataRetryPolicy retryPolicy;
+    private final SdkAsyncHttpClient httpClient;
+    private final EndpointMode endpointMode;
+    private final URI endpoint;
+    private final RequestMarshaller requestMarshaller;
+    private final Duration tokenTtl;
+    private final ScheduledExecutorService asyncRetryScheduler;
+
+    private DefaultEc2MetadataAsyncClient(Ec2MetadataAsyncBuilder builder) {
+        this.retryPolicy = builder.retryPolicy != null ? builder.retryPolicy
+                                                       : Ec2MetadataRetryPolicy.builder().build();
+        this.httpClient = builder.httpClient != null ? builder.httpClient
+                                                     : AwsCrtAsyncHttpClient.create();
```

Comment:
That is actually a very good point. It was not discussed, no, which async http should be used by default. Whichever one we used, it will have to be included in the classpath of the user, indeed.

We make sure users have the CRT on their classpath automatically by adding the dependency to the `aws-crt-client` in the `pom.xml` of the IMDS module. Right now, users that want to use another http-client would have to include that client in their `pom.xml` (and optionally exclude the `aws-crt-client` dependency)

Reply:
So far we've avoided forcing users into taking a dependency on the CRT without them making an active choice, except for with Transfer Manager, because we don't own CRT and there may be issues (I'm vague because it's been a while). In S3 for instance, which uses auth-crt, we check the classpath to see if it's included. Let's discuss offline what the best course of action is.

Reply:
as discussed, we will instead use the `DefaultSdkHttpClientBuilder` to instanciate the http client